wallet: allowing walletpassphrase timeout to be MAX_SLEEP_TIME if set to -1

kevkevinpal opened the pull request:
In this change we allow the timeout for walletpassphrase to be
set to MAX_SLEEP_TIME,
if set as -1 then we then use the MAX_SLEEP_TIME amount

context from PR 28403

added release notes for RPC Wallet

Also tests were modified to use the max timeout amount using timeout = -1

<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

kevkevinpal committed:
wallet: walletpassphrase timeout to be set to MAX_SLEEP_TIME

In this change we allow the timeout for walletpassphrase to be
set to MAX_SLEEP_TIME,
if set as -1 then we then use the MAX_SLEEP_TIME amount

context from PR 28403

added release notes for RPC Wallet

Also tests were modified to use the max timeout amount using timeout = -1

kristapsk commented:
Concept NACK. I don't think this is a good idea for mainnet.

kevkevinpal commented:
> Concept NACK. I don't think this is a good idea for mainnet.

Any specific reason why? A user can currently input a time larger than MAX_SLEEP_TIME and it would be set to it, we are just defaulting to it if no arg is added.

Or are you more concerned that a user wouldn't notice the optional arg and unexpectedly have a command with a long timeout running?

There was also the suggestion of 0 being used instead of omitting the arg to signal max timeout

kristapsk commented:
> Or are you more concerned that a user wouldn't notice the optional arg and unexpectedly have a command with a long timeout running?

Yes

kevkevinpal commented:
@kristapsk I modified the PR to [1cc6b1c](https://github.com/bitcoin/bitcoin/pull/28454/commits/1cc6b1cf6ba5ccef9b1c5716295b4b14cf400250) which now keeps it as a non optional param but now if the user sets 0 then we will use the MAX_SLEEP_TIME

ishaanam reviewed (commented):
This looks good to me, though I think it could use some release notes. An alternative to this could be that if the timeout is set to 0, then the wallet never automatically locks. I've implemented this here (da409dc5df6bb2ef3208b39d0b2a563debd03376), if anyone is curious, but the approach implemented in this PR is simpler.

luke-jr commented:
>(If someone wanted to immediately re-lock, it seems easier to do it synchronously with the walletlock RPC, than to wait for the async callback and poll the locked-state in a loop)

I wonder if there's an argument around a JSON-RPC batch that unlocks, runs some method, and re-locks... Probably wouldn't work today, but might be worth adding?

kevkevinpal commented:
forced pushed to [7deb571](https://github.com/bitcoin/bitcoin/pull/28454/commits/7deb571196230090f9a78589c439b36020431f47)

added release notes and changed tests in this commit to use timeout=0

kevkevinpal commented:
> > That said, I think this breaks https://en.wikipedia.org/wiki/Principle_of_least_astonishment (_should behave in a way that most users will expect it to behave_), and in a less secure direction.
> 
> The same is used for the rpc client/server timeout, so I think it seems more consistent to map `0` to max. Though, if it is too controversial, any negative value can be mapped to max instead?

I would be open to using a negative value eg(-1) since any negative value rn throws an error. But I do think if we're using 0 else where we should either continue doing the same or change it there aswell

jonatack reviewed (commented):
What would be a use case for encrypting a wallet and then setting a 3-year timeout?  Maybe I'm missing something, but it seems like doing so should be explicit (as it is currently).  I would expect users to pass as short a value as possible, i.e. a few seconds or minutes.

kevkevinpal commented:
Thanks @jonatack I added those changes and pushed [a7bc5db](https://github.com/bitcoin/bitcoin/pull/28454/commits/a7bc5db7526387248c798f337f952c47e07676aa)

kevkevinpal commented:
> @kevkevinpal I think the OP is out of date.

Sorry not sure what you mean here

ishaanam commented:
@kevkevinpal 
The description of this PR no longer describes what has been implemented because this PR was changed to not allow `timeout` to be optional.

kevkevinpal commented:
> @kevkevinpal The description of this PR no longer describes what has been implemented because this PR was changed to not allow `timeout` to be optional.

Thanks, updated the PR description to match the commit message

ishaanam commented:
I think that it would also be fine to only implement this for testnet and regtest. The wallet locking too soon has been a frequently occurring problem with wallet tests. If this is required for the batch feature however, then that is another case.

kevkevinpal commented:
> I think that it would also be fine to only implement this for testnet and regtest. The wallet locking too soon has been a frequently occurring problem with wallet tests. If this is required for the batch feature however, then that is another case.

Added if block to check if non-mainnet (let me know if you want to exclude signet for any reason). 
Also updated the help doc to say that this would only happen in non mainnet.
Updated commit message to include info about this only working in non mainnet in [c4ce733](https://github.com/bitcoin/bitcoin/pull/28454/commits/c4ce733976e08a3be89fc3e756021f22e76ad484)

kevkevinpal commented on src/wallet/rpc/encrypt.cpp:
```diff
@@ -59,6 +60,12 @@ RPCHelpMan walletpassphrase()
         constexpr int64_t MAX_SLEEP_TIME = 100000000; // larger values trigger a macos/libevent bug?
+
+        // Allow sleep time to be set to max time by using zero in non mainnet
+        if (Params().GetChainType() != ChainType::MAIN && nSleepTime == 0) {
```
I agreed I can switch it back to being for all chaintypes

kevkevinpal commented:
> On a second thought, I wonder if someone uses `0` as a brittle alias for `walletlock`. Maybe not worth it to risk breaking that?

I can switch it from 0 to -1 since negative values are not being used currently for `timeout`

kevkevinpal commented:
Removed non mainnet if block and switched to using -1 instead of 0 to set time as `MAX_SLEEP_TIME` in [0fc8db0](https://github.com/bitcoin/bitcoin/pull/28454/commits/0fc8db0e2e823504224eda57a3742b2214eddab8)

kevkevinpal commented:
rebased to [76b1c4c](https://github.com/bitcoin/bitcoin/pull/28454/commits/76b1c4ca6eea327328c7a0b035e6f2b1a2211e51) and since this [PR 28617](https://github.com/bitcoin/bitcoin/pull/28617) has been merged only the test-framework will be modified so I am not sure if this change is needed anymore. Thoughs @maflcko, I am happy to close this

kevkevinpal commented:
> rebased to [76b1c4c](https://github.com/bitcoin/bitcoin/pull/28454/commits/76b1c4ca6eea327328c7a0b035e6f2b1a2211e51) and since this [PR 28617](https://github.com/bitcoin/bitcoin/pull/28617) has been merged only the test-framework will be modified so I am not sure if this change is needed anymore.
> 
> Thoughts @maflcko? I am happy to close this

closing this pull request
